core/state/statedb.go

```diff
@@ -47,24 +47,25 @@ type revision struct {
 	journalIndex int
 }
 
-// StateDB structs within the ethereum protocol are used to store anything
-// within the merkle trie. StateDBs take care of caching and storing
-// nested states. It's the general query interface to retrieve:
-//
-// * Contracts
-// * Accounts
-//
-// Once the state is committed, tries cached in stateDB (including account
-// trie, storage tries) will no longer be functional. A new state instance
-// must be created with new root and updated database for accessing post-
-// commit states.
+// StateDB structs within the ethereum protocol are used to read, write and
+// hash ethereum states.
 type StateDB struct {
-	db         Database
+	db Database
+
+	// This reader is associated with a specific state and gets re-constructed
+	// with a new state root after a commit operation.
+	reader StateReader
+
+	// This tool preloads essential trie elements and caches them within
+	// the trie structure, reducing the time required for node resolution
+	// in the late trie operation. It's optional and should be configured
+	// explicitly.
 	prefetcher *triePrefetcher
-	trie       Trie
-	hasher     crypto.KeccakState
-	snaps      *snapshot.Tree    // Nil if snapshot is not available
-	snap       snapshot.Snapshot // Nil if snapshot is not available
+
+	// primaryTrie represents the main state trie, initially nil and set upon
+	// first access. It's reset to nil when stateDB is committed to ensure a
+	// reload with the new root after the commit.
+	trie Trie
```

Comment:
Regarding the state.Database, my approach is to maintain it as stateless. Since the stateDB supports Copy functionality, introducing a Copy function for state.Database is necessary and over-complicated if it's stateful.

On the other hand, when it comes to the state.Reader, it operates with a stateful design. It efficiently caches pre-loaded items to enhance the speed of subsequent state reads. However, it's important to note that this interface doesn't have the "tree" concept; instead, it hides all the background implementation details but just provide the access to account and storage slots. So tree is certainly not a part of state.Reader.